m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,831 @@
+                    String idTagValue = String.format(bucketIdFmt, i);
+
+                    tags.put(bucketIdTagName, idTagValue);
+                    tags.put(bucketTagName,
+                        String.format("%s-%s",
+                            durationBucketString(bucketPairs[i].lowerBoundDuration()),
+                            durationBucketString(bucketPairs[i].upperBoundDuration())
+                        )
+                    );
+
+                    break;
+                }
+            }
+
+            reportCounterInternal(name, tags, samples);
+        }
+    }
+
+    // Relies on the calling function to provide guarantees of the reporter being open
+    private void reportCounterInternal(String name, Map<String, String> tags, long value) {
+        CountValue countValue = new CountValue();
+        countValue.setI64Value(value);
+
+        MetricValue metricValue = new MetricValue();
+        metricValue.setCount(countValue);
+
+        Metric metric = newMetric(name, tags, metricValue);
+
+        queueSizedMetric(new SizedMetric(metric, calculateSize(metric)));
+    }
+
+    private Metric newMetric(String name, Map<String, String> tags, MetricValue metricValue) {
+        Metric metric = new Metric(name);
+        metric.setTags(toMetricTagSet(tags));
+        metric.setTimestamp(System.currentTimeMillis() * Duration.NANOS_PER_MILLI);
+        metric.setMetricValue(metricValue);
+
+        return metric;
+    }
+
+    private void queueSizedMetric(SizedMetric sizedMetric) {
+        try {
+            metricQueue.put(sizedMetric);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+
+            LOG.warn(String.format("Interrupted queueing metric: %s", sizedMetric.getMetric().getName()));
+        }
+    }
+
+    private class Processor implements Runnable {
+        List<Metric> metrics = new ArrayList<>(freeBytes / 10);
+        int metricsSize = 0;
+
+        @Override
+        public void run() {
+            try {
+                while (!executor.isShutdown()) {
+                    // This `take` call will block until there is an item on the queue to take.
+                    // When this reporter is closed, shutdownNow will be called on the executor,
+                    // which will interrupt this thread and proceed to the `InterruptedException`
+                    // catch block.
+                    SizedMetric sizedMetric = metricQueue.take();
+
+                    flushMetricIteration(sizedMetric);
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            } finally {
+                flushRemainingMetrics();
+
+                // Always arrive at phaser to prevent deadlock
+                phaser.arrive();
+            }
+        }
+
+        private void flushMetricIteration(SizedMetric sizedMetric) {
```

Comment:
There's a subtle bug in this implementation: consider the case where a user emits a single metric very infrequently. You'd wait for a long time (till the size was reached), and/or the scope was closed, before flushing a metric. 

We should be flushing after a defined period regardless of the number of metrics queued.

Reply:
Is this still applicable?

Reply:
Forgot to reply to this one. I made the change [here](https://github.com/uber-java/tally/pull/12/files#diff-0ab29d028dfc2be47a19838ce2522eb0R573) to replace `take` with `poll`, which returns null if it doesn't get anything after the specified duration. I then handle the null case by flushing what we currently have.

Reply:
Cool, could you add tests for these cases